tapeagents/orchestrator.py

```diff
@@ -147,6 +167,119 @@ def _implementation():
+        yield MainLoopEvent(agent_tape=agent_tape)
+
+        # --- RUN THE ENVIRONMENT ---
+        if any([isinstance(step, StopStep) for step in agent_tape.steps]):
+            logger.info(f"Agent emitted final step {agent_tape.steps[-1]}")
+            yield MainLoopEvent(status=MainLoopStatus.FINISHED)
+            return
+        try:
+            tape = await environment.areact(agent_tape)
+        except NoActionsToReactTo:
+            yield MainLoopEvent(status=MainLoopStatus.NO_ACTIONS)
+            return
+        except ExternalObservationNeeded:
+            yield MainLoopEvent(status=MainLoopStatus.EXTERNAL_INPUT_NEEDED)
+            return
+        for observation in tape[len(agent_tape) :]:
+            logger.info(colored(f"ENV: {step_view(observation, trim=True)}", "yellow"))
+            yield MainLoopEvent(observation=observation)
+        yield MainLoopEvent[TapeType](env_tape=tape)
+
+        # --- REPEAT ---
+        n_loops += 1
+
+
+async def execute_agent(
+    agent: Agent[TapeType],
+    start_tape: TapeType,
+    environment: AsyncEnvironment,
+    session: aiohttp.ClientSession,
+    max_loops: int = 50,
+) -> TapeType:
+    final_tape = start_tape
+    try:
+        async for event in async_main_loop(agent, start_tape, environment, session, max_loops):
+            if event.agent_event and event.agent_event.final_tape:
+                final_tape = event.agent_event.final_tape
+            elif event.env_tape:
+                final_tape = event.env_tape
+    except Exception as e:
+        final_tape.metadata.error = f"Agent loop exception: {e}"
+        logger.error(colored(f"Agent loop exception: {e}, stopping", "red"))
+    final_tape.metadata.parent_id = start_tape.metadata.id
+    return final_tape
+
+
+class EnvironmentManager:
+    def __init__(self, cfg: DictConfig, n_envs: int = 1):
+        self.cfg = cfg
+        self.semaphore = asyncio.Semaphore(n_envs)
+        self.envs: list[ToolCollectionEnvironment] = [instantiate(cfg) for _ in range(n_envs)]
+        self.exit_stack = AsyncExitStack()
+        logger.info(f"Created {len(self.envs)} environments")
+
+    async def __aenter__(self):
+        for env in self.envs:
+            await env.ainitialize()
+            self.exit_stack.push_async_callback(env.aclose)
+        logger.info(f"Initialized {len(self.envs)} environments")
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        await self.exit_stack.aclose()
+
+    @asynccontextmanager
+    async def get_env(self):
+        logger.info("waiting for env")
+        async with self.semaphore:
+            env = self.envs.pop(0)
+            logger.info(f"pop env, {len(self.envs)} environments left")
+            try:
+                yield env
+            finally:
+                await env.areset()
+                self.envs.append(env)
+                logger.info(f"push env back, {len(self.envs)} environments available")
+
+
+async def execute_with_env(
```

Comment:
Maybe `get_env_and_execute`? Not sure how help this utility is, might be better to write this code explicitly everywhere.